### DISCUSSION TITLE
DeepTrack2 2.0.0

### DISCUSSION THREAD

**Discussion by giovannivolpe** [DeepTrackAI/DeepTrack2#205] [Announcements]
Stable release of DeepTrack2 2.0.0
Release date: 19 March 2023

PyPi 2.0.0
https://pypi.org/project/deeptrack/2.0.0/

This major release is a significant update that includes the migration to PyTorch and integration with the `deeplay` library. This version focuses on PyTorch/Deeplay integration, performance optimization, and a more rigorous approach to utilizing static data from disk.

### Image Objects
- Pipelines no longer return `Image` objects by default, to improve performance and compatibility with other libraries. This change will not affect you unless you directly interact with the `.properties` attribute or call `.get_properties(.)` on pipeline outputs.
- If you need `Image` objects, you can enable them by using `pipeline.store_properties()`, restoring the behavior from previous releases.

### TensorFlow Dependency Removed
- TensorFlow has been removed as a dependency due to compatibility issues and the complexity of maintaining it alongside major Python versions, especially on Windows.
- We will not actively support TensorFlow versions newer than 2.10. Users are encouraged to manage TensorFlow installations independently.

### Global Changes
- Many submodules are now lazily loaded, only initializing when needed. This change significantly speeds up initial load times and reduces TensorFlow-related issues for users not requiring neural network functionalities.

### Sources
- Introducing `Sources`, a new method for efficiently operating on static datasets. This feature addresses common problems with data manipulation and pipeline evaluation, allowing for more functional and efficient dataset handling.

### PyTorch Integration
- PyTorch code is available in the `pytorch` submodule, which is lazy-loaded to prevent unnecessary overhead. This update includes:
  - `pytorch.Dataset`: A subclass of `torch.utils.data.Dataset`, making DeepTrack pipelines compatible with standard `DataLoader`s.
  - `pytorch.ToTensor`: A feature for converting pipeline outputs to PyTorch tensors, with support for specifying the `dtype`.

### Performance Optimization
- This release is centered on performance optimization and a more effective way of using static data, enhancing the user experience and streamlining workflows.

## `deeplay`
- The `deeplay` library is now accessible as `deeptrack.deeplay`, offering a powerful and flexible foundation for constructing neural networks with PyTorch.

This update marks a significant milestone in DeepTrack's development, emphasizing our commitment to performance, usability, and modern deep learning practices. We're excited to see how these improvements will enable our users to achieve even more with DeepTrack.

Cite as:
Benjamin Midtvedt, Saga Helgadottir, Aykut Argun, Jesús Pineda, Daniel Midtvedt, Giovanni Volpe.
"Quantitative Digital Microscopy with Deep Learning."
Applied Physics Reviews 8, 011310 (2021)
https://doi.org/10.1063/5.0034891

<hr /><em>This discussion was created from the release <a href='https://github.com/DeepTrackAI/DeepTrack2/releases/tag/2.0.0'>DeepTrack2 2.0.0</a>.</em>

**Comment by b-grimaud** [comment]
Hi,
I understand this is a pre-release, especially since PyPI still shows 1.7.0.
Does it involve API changes significant enough that the docs and notebook examples currently available are obsolete ?
I also looked for any specific instructions on how to build from source but the links to the docs in the readme is apparently dead.

**Comment by BenjaminMidtvedt** [comment]
@BptGrm Hello! You can install the pre-release with `pip install --pre deeptrack`

The API changes are not very significant compared to the `1.7.0` release, but not all examples have been updated to work with the 2.0.0 version (hence the pre-release)
